src/crimson/os/seastore/async_cleaner.cc

```diff
@@ -1168,88 +1168,91 @@ SegmentCleaner::clean_space_ret SegmentCleaner::clean_space()
-      reclaim_state->end_pos
-    ).si_then([this, &t](auto pin_list) {
-      if (!pin_list.empty()) {
-	auto it = pin_list.begin();
-	auto &first_pin = *it;
-	if (first_pin->get_key() < reclaim_state->start_pos) {
-	  // BackrefManager::get_mappings may include a entry before
-	  // reclaim_state->start_pos, which is semantically inconsistent
-	  // with the requirements of the cleaner
-	  pin_list.erase(it);
-	}
-      }
-      return backref_manager.retrieve_backref_extents_in_range(
-        t,
-        reclaim_state->start_pos,
-        reclaim_state->end_pos
-      ).si_then([pin_list=std::move(pin_list)](auto extents) mutable {
-        return std::make_pair(std::move(extents), std::move(pin_list));
-      });
-    });
-  }).safe_then([this, FNAME, pavail_ratio, start](auto weak_read_ret) {
-    return seastar::do_with(
-      std::move(weak_read_ret.first),
-      std::move(weak_read_ret.second),
-      (size_t)0,
-      (size_t)0,
-      [this, FNAME, pavail_ratio, start](
-        auto &backref_extents, auto &pin_list, auto &reclaimed, auto &runs)
-    {
-      return do_reclaim_space(
-          backref_extents,
-          pin_list,
-          reclaimed,
-          runs
-      ).safe_then([this, FNAME, pavail_ratio, start, &reclaimed, &runs] {
-        stats.reclaiming_bytes += reclaimed;
-        auto d = seastar::lowres_system_clock::now() - start;
-        DEBUG("duration: {}, pavail_ratio before: {}, repeats: {}",
-              d, pavail_ratio, runs);
-        if (reclaim_state->is_complete()) {
-          auto segment_to_release = reclaim_state->get_segment_id();
-          INFO("reclaim {} finish, reclaimed alive/total={}",
-               segment_to_release,
-               stats.reclaiming_bytes/(double)segments.get_segment_size());
-          stats.reclaimed_bytes += stats.reclaiming_bytes;
-          stats.reclaimed_segment_bytes += segments.get_segment_size();
-          stats.reclaiming_bytes = 0;
-          reclaim_state.reset();
-          return sm_group->release_segment(segment_to_release
-          ).handle_error(
-            clean_space_ertr::pass_further{},
-            crimson::ct_error::assert_all{
-              "SegmentCleaner::clean_space encountered invalid error in release_segment"
-            }
-          ).safe_then([this, FNAME, segment_to_release] {
-            auto old_usage = calc_utilization(segment_to_release);
-            if(unlikely(old_usage != 0)) {
-              space_tracker->dump_usage(segment_to_release);
-              ERROR("segment {} old_usage {} != 0",
-                     segment_to_release, old_usage);
-              ceph_abort();
+  return seastar::do_with(
+    std::vector<CachedExtentRef>(),
+    backref_pin_list_t(),
+    [this, FNAME, pavail_ratio, start](auto& backref_extents, auto& pin_list) {
+    return repeat_eagain([this, &backref_extents, &pin_list] {
+      return extent_callback->with_transaction_intr(
+        Transaction::src_t::READ,
```

Comment:
Why this commit is necessary to allow conflicts on the "retrieve_from_backref_tree" transaction?

Why it is possible to conflict now, due to parallel calls to `SegmentCleaner::clean_space()`?
https://github.com/ceph/ceph/blob/f70ba291fdd38d6cfb967dcb40cd9f3a87f97c36/src/crimson/os/seastore/extent_placement_manager.cc#L634-L662

Reply:
> due to parallel calls to `SegmentCleaner::clean_space()`?

Yes, "CLEANER_MAIN" and "CLEANRE_COLD" might conflict on the backref extents. Such extents located in the reclaimed range are going to be rewritten by "CLEANER_MAIN" and "CLEANER_COLD" alose needs these backref extents to retrieve live extents on the cold tier.

Reply:
Please refer to https://github.com/ceph/ceph/pull/50191#discussion_r1113865629